src/monodroid/jni/monodroid-glue.c

```diff
@@ -647,11 +666,16 @@ get_libmonosgen_path ()
 	copy_monosgen_to_internal_location (primary_override_dir, external_override_dir);
 	copy_monosgen_to_internal_location (primary_override_dir, external_legacy_override_dir);
 
-	int i;
-	for (i = 0; i < MAX_OVERRIDES; ++i)
-		TRY_LIBMONOSGEN (override_dirs [i])
+	if (!embedded_dso_mode) {
+		for (i = 0; i < MAX_OVERRIDES; ++i)
+			TRY_LIBMONOSGEN (override_dirs [i]);
+	}
 #endif
-	TRY_LIBMONOSGEN (app_libdir)
+	if (!embedded_dso_mode) {
```

Comment:
This condition concerns me, in that it means that if a *Debug* build was using "embedded DSO support," that Debug build can't use Improved Fast Deployment, because we only look in the override directories when `embedded_dso_mode`==0.

Reply:
No, this is incorrect. We look in override dirs in embedded mode. See the following:

https://github.com/grendello/xamarin-android/blob/issue1906/src/monodroid/jni/monodroid-glue.c#L890
https://github.com/grendello/xamarin-android/blob/issue1906/src/monodroid/jni/monodroid-glue.c#L3278
https://github.com/grendello/xamarin-android/blob/issue1906/src/monodroid/jni/monodroid-glue.c#L3293
https://github.com/grendello/xamarin-android/blob/issue1906/src/monodroid/jni/monodroid-glue.c#L3494
https://github.com/grendello/xamarin-android/blob/issue1906/src/monodroid/jni/monodroid-glue.c#L3996

I tested this PR in Debug mode :)

Reply:
> I tested this PR in Debug mode :)

Which is wonderful!  The problem is that my "static analysis" (trying to read this PR) doesn't *easily* see *why* that works, not easily, not obviously.  Here, within `get_libmonosgen_path()`, when `embedded_dso_mode==TRUE`, *how* do we find `.__override__/.../libmonosgen-2.0.so`?  *Do we* find `libmonosgen-2.0.so` within this function at all, or is it found "elsewhere"?  If it is "elsewhere," *where* is that "elsewhere"?

Turns out, that "elsewhere" is in `Java_mono_android_Runtime_init()`:

	if (embedded_dso_mode) {
 		libmonosgen_handle = load_dso_from_any_directories (MONO_SGEN_SO, sgen_dlopen_flags);
 	}

I wonder if things would be clear if instead of a `get_libmonosgen_path()` we had a `load_libmonosgen()`, so that the "find `libmonosgen`" and "load `libmonosgen`" logic weren't so far apart. :-/

Reply:
Yes, that would be better - this code could use a lot of improvements all over the map. However, I don't want to muddle the actual topic of this particular PR with logic and code cleanup. I'd like to make such changes/improvements **after** we merge in the C++ PR. I purposefully stayed as close to original code when implementing this feature as possible (even though I often felt like rewriting large portions of what I saw - but that would mean going down the rabbit hole and making this PR something that cannot be properly reviewed)